linalg/batchlinalg.cpp

```diff
@@ -0,0 +1,391 @@
+// Copyright (c) 2010-2024, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+
+// Implementation of batchlinalg class
+
+#include "batchlinalg.hpp"
+#include "../general/forall.hpp"
+
+#if defined(MFEM_USE_UMPIRE)
+#define MFEM_RELEASE_MEM(o) o.GetMemory().DeleteDevice(false)
+#else
+#define MFEM_RELEASE_MEM(o)
+#endif
+
+#if defined(MFEM_USE_CUDA)
+#include <cublas_v2.h>
+#include <cusolverDn.h>
+#define MFEM_cu_or_hip(stub) cu##stub
+#define MFEM_Cu_or_Hip(stub) Cu##stub
+#define MFEM_CU_or_HIP(stub) CU##stub
+#define MFEM_CUDA_or_HIP(stub) CUDA##stub
+#elif defined(MFEM_USE_HIP)
+#include <hipblas/hipblas.h>
+#include <hip/hip_runtime.h>
+#include <hip/hip_runtime_api.h>
+#define MFEM_cu_or_hip(stub) hip##stub
+#define MFEM_Cu_or_Hip(stub) Hip##stub
+#define MFEM_CU_or_HIP(stub) HIP##stub
+#define MFEM_CUDA_or_HIP(stub) HIP##stub
+#else
+#define MFEM_cu_or_hip(stub)
+#define MFEM_Cu_or_Hip(stub)
+#define MFEM_CU_or_HIP(stub)
+#define MFEM_CUDA_or_HIP(stub)
```

Comment:
we have these defined in a few places now, do you want to move them to a shared header? Or should we do that in a followup ?

Reply:
I think now would be good to do